Fix discovery cli to print devices not printed during discovery timeout

sdb9696 opened the pull request:
Bug identified while testing the [discovery sleep PR](https://github.com/python-kasa/python-kasa/pull/656).  If discovery times out before it can print all the devices they are under reported.

Also identified issues with exceptions in cli `state` called from discovery not being retrieved which is why tests did not pick up this bug.

Reduces `test_discovery` duration from 34s to 13s.

Fixes #672

rytilahti commented on kasa/device_factory.py:
```diff
@@ -132,7 +132,7 @@ def get_device_class_from_family(device_type: str) -> Optional[Type[SmartDevice]
         "SMART.TAPOBULB": TapoBulb,
         "SMART.KASAPLUG": TapoPlug,
-        "SMART.KASASWITCH": TapoBulb,
+        "SMART.KASASWITCH": TapoPlug,
```
To my understanding this would break the brightness functionality, so I'd wait until we get a more modular base class for the SMART devices.

sdb9696 replied:
Reverted

rytilahti commented on kasa/cli.py:
```diff
@@ -440,6 +440,23 @@ async def print_discovered(dev: SmartDevice):
+
+    for device in discovered_devices.values():
+        await device.protocol.close()
+        # TODO transports need to permit close to avoid
+        # ERROR:asyncio:Unclosed client session
```
I think we will want to fix this properly, so I'll mark this PR as a draft to make it easier to keep track on things that are ready to be merged.

rytilahti reviewed (approved):
I'm wondering if there's no better way than waiting for the discover_task to get initialized in a busy loop, but otherwise this looks good to me. Would you mind changing the container for unsupported in the same PR?

sdb9696 commented:
> I'm wondering if there's no better way than waiting for the discover_task to get initialized in a busy loop, but otherwise this looks good to me. Would you mind changing the container for unsupported in the same PR?

Now updated to use `asyncio.event.wait` instead of the busy loop.

Changed the container for `auth_failed` back to a list as it didn't need to be a dict so it's now consistent with `unsupported`

sdb9696 commented on kasa/cli.py:
```diff
@@ -411,7 +411,7 @@ async def discover(ctx):
     sem = asyncio.Semaphore()
     discovered = dict()
     unsupported = []
```
Changed the container for `auth_failed` back to a list as it didn't need to be a dict so it's now consistent with `unsupported` and unchanged in this pr.
